### PR TITLE
Feature/csi 3984 version in volume name

### PR DESCRIPTION
**Pull request by ArbelNathan**

### CODE REVIEW

File: controller/controller_server/csi_controller_server.py [hunk 1]
```diff
@@ -514,8 +514,15 @@ def _get_snapshot_final_name(self, volume_parameters, name, array_mediator):
                                            config.SNAPSHOT_TYPE_NAME)
         return name
 
+    def _get_version_letter(self):
+        version = self.get_identity_config("version")
+        _, minor_version, _ = version.split(config.VERSION_DELIMITER)
```
Comment (oriyarde) [inline controller/controller_server/csi_controller_server.py:519]: [`from packaging.version import Version`](https://github.com/IBM/ibm-block-csi-driver/blob/20605cc8ce2249e77f3b6d71b6545ed9928bccf1/controller/array_action/array_mediator_svc.py#L5)
```suggestion
        raw_version = self.get_identity_config("version")
        version = Version(raw_version)
        minor_version = version.minor
```

File: controller/controller_server/config.py [hunk 2]
```diff
@@ -15,6 +16,8 @@
 VOLUME_WWN_LENGTH = 32
 MAX_ARRAY_TYPE_LENGTH = max(map(len, ALL_ARRAY_TYPES))
 NUMBER_OF_DELIMITERS_IN_VOLUME_ID = 2
+NUMBER_OF_LETTERS = 26
+FIRST_LETTER = 'a'
```
Comment (oriyarde) [inline controller/controller_server/config.py:20]:
```
from string import ascii_lowercase
```
```suggestion
NUMBER_OF_LETTERS = len(ascii_lowercase)
FIRST_LETTER = ascii_lowercase[0]
```

File: controller/controller_server/csi_controller_server.py [hunk 3]
```diff
@@ -514,8 +515,16 @@ def _get_snapshot_final_name(self, volume_parameters, name, array_mediator):
                                            config.SNAPSHOT_TYPE_NAME)
         return name
 
+    def _get_version_letter(self):
+        raw_version = self.get_identity_config("version")
+        version = Version(raw_version)
+        minor_version = version.minor
+        minor_version_in_letters_range = int(minor_version) % config.NUMBER_OF_LETTERS
```
Comment (oriyarde) [inline controller/controller_server/csi_controller_server.py:522]:
```suggestion
        minor_version_offset_in_letters_range = int(minor_version) % config.NUMBER_OF_LETTERS
```

File: controller/controller_server/csi_controller_server.py [hunk 4]
```diff
@@ -514,8 +515,16 @@ def _get_snapshot_final_name(self, volume_parameters, name, array_mediator):
...
+        base_value = ord(config.FIRST_LETTER)
```
Comment (oriyarde) [inline controller/controller_server/csi_controller_server.py:523]:
```suggestion
        base_ordinal = ord(config.FIRST_LETTER)
```
```suggestion
        letters_range_start = ord(config.FIRST_LETTER)
```

Comment (ArbelNathan) [reply controller/controller_server/csi_controller_server.py]: done

File: controller/controller_server/csi_controller_server.py (hunk 3 above)
Comment (ArbelNathan) [reply controller/controller_server/csi_controller_server.py]: done

File: controller/controller_server/config.py (hunk 2 above)
Comment (ArbelNathan) [reply controller/controller_server/config.py]: done, nice!

File: controller/controller_server/csi_controller_server.py (hunk 1 above)
Comment (ArbelNathan) [reply controller/controller_server/csi_controller_server.py]: done

Comment (ArbelNathan) [reply controller/controller_server/csi_controller_server.py]: done

File: controller/tests/controller_server/csi_controller_server_test.py [hunk 5]
```diff
@@ -612,36 +613,60 @@ def test_create_volume_with_other_exception(self):
...
         self.mediator.create_volume.return_value = utils.get_mock_mediator_response_volume(10, "volume", "wwn", "xiv")
         self.mediator.validate_supported_space_efficiency = Mock()
         self.servicer.CreateVolume(self.request, self.context)
         self.assertEqual(self.context.code, grpc.StatusCode.OK)
-        self.mediator.create_volume.assert_called_once_with(final_name, 10, space_efficiency, POOL, None)
+        if not prefix:
+            prefix = DEFAULT_OBJECT_PREFIX
+        call_args = self.mediator.create_volume.call_args
+        volume_final_name, _, space_efficiency, _, _ = call_args[0]
+        self.assertIn(prefix, volume_final_name)
```
Comment (oriyarde) [inline controller/tests/controller_server/csi_controller_server_test.py:628]:
```suggestion
        self.assertTrue(volume_final_name.startwith(prefix))

```

File: controller/tests/controller_server/csi_controller_server_test.py [hunk 6]
```diff
@@ -612,36 +613,60 @@ def test_create_volume_with_other_exception(self):
         self.create_volume_returns_error(return_code=grpc.StatusCode.INTERNAL,
                                          err=Exception("error"))
 
-    def _test_create_volume_parameters(self, final_name="default_some_name", space_efficiency=None):
-        self.mediator.default_object_prefix = "default"
-        self.request.name = "some_name"
+    def _test_create_volume_parameters(self, prefix="", expected_space_efficiency=None):
```
Comment (oriyarde) [inline controller/tests/controller_server/csi_controller_server_test.py:616]:
```suggestion
    def _test_create_volume_parameters(self, expected_prefix="", expected_space_efficiency=None):

```

File: controller/tests/controller_server/csi_controller_server_test.py [hunk 7]
```diff
@@ -323,18 +328,15 @@ def test_create_snapshot_with_other_exception(self):
...
+        self._prepare_create_snapshot_mocks(storage_agent)
+        self.request.parameters[config.PARAMETERS_SNAPSHOT_NAME_PREFIX] = OBJECT_PREFIX
 
         self.servicer.CreateSnapshot(self.request, self.context)
 
         self.assertEqual(self.context.code, grpc.StatusCode.OK)
-        self.mediator.create_snapshot.assert_called_once_with(SNAPSHOT_VOLUME_WWN, "prefix_some_name", None, None)
+        call_args = self.mediator.create_snapshot.call_args
+        _, snapshot_final_name, _, _ = call_args[0]
+        self.assertIn(OBJECT_PREFIX, snapshot_final_name)
```
Comment (oriyarde) [inline controller/tests/controller_server/csi_controller_server_test.py:339]: I think we need to implement `self.assertStartsWith`

File: controller/controller_server/csi_controller_server.py [hunk 8]
```diff
@@ -526,18 +535,17 @@ def _get_object_final_name(self, volume_parameters, name, array_mediator, object
...
-            full_name = self._join_object_prefix_with_name(prefix, hashed_name)
+            full_name = self._join_name_parts(prefix, hashed_name)
         return full_name[:array_mediator.max_object_name_length]
 
-    def _join_object_prefix_with_name(self, prefix, name):
-        if prefix:
-            return settings.NAME_PREFIX_SEPARATOR.join((prefix, name))
-        return name
+    def _join_name_parts(self, prefix, name, version=''):
+        return settings.NAME_PREFIX_SEPARATOR.join(filter(None, (prefix, version, name)))
```
Comment (oriyarde) [inline controller/controller_server/csi_controller_server.py:548]:
```suggestion
        return settings.NAME_PARTS_SEPARATOR.join(filter(None, (prefix, version, name)))
```

File: controller/controller_server/csi_controller_server.py [hunk 9]
```diff
@@ -526,18 +535,17 @@ def _get_object_final_name(self, volume_parameters, name, array_mediator, object
...
+        version_letter = self._get_version_letter()
+        full_name = self._join_name_parts(prefix, name, version_letter)
         if len(full_name) > array_mediator.max_object_name_length:
             hashed_name = utils.hash_string(name)
-            full_name = self._join_object_prefix_with_name(prefix, hashed_name)
+            full_name = self._join_name_parts(prefix, hashed_name)
         return full_name[:array_mediator.max_object_name_length]
 
-    def _join_object_prefix_with_name(self, prefix, name):
-        if prefix:
```
Comment (oriyarde) [inline controller/controller_server/csi_controller_server.py:538]: what if the prefix is `None` (e.g. ds8k)? was this change even tested?

File: controller/tests/controller_server/csi_controller_server_test.py [hunk 10]
```diff
@@ -612,36 +613,60 @@ def test_create_volume_with_other_exception(self):
...
         self.request.parameters[config.PARAMETERS_VOLUME_NAME_PREFIX] = ""
         self._test_create_volume_parameters()
 
-    def _test_create_volume_with_parameters_by_system_prefix(self, get_array_connection_info_from_secrets, prefix,
-                                                             final_name="default_some_name",
+    @patch("controller.controller_server.csi_controller_server.get_agent")
+    def test_create_volume_with_short_name_length(self, storage_agent):
+        storage_agent.return_value = self.storage_agent
+        self.mediator.max_object_name_length = 16
+        self._test_create_volume_parameters(prefix="p" * 5)
```
Comment (oriyarde) [inline controller/tests/controller_server/csi_controller_server_test.py:646]: I think we need to control the max prefix length in the test class init

**Pull request closed without merging**